rust/operator-binary/src/controller.rs

```diff
@@ -0,0 +1,1002 @@
+    {
+        true => Ok(false),
+        false => {
+            if !sensitive_config.auto_generate {
+                return Err(Error::SensitiveKeySecretMissing {name: sensitive_config.key_secret.clone(), namespace: namespace.to_string()});
+            }
+            tracing::info!("No existing sensitive properties key found, generating new one");
+            let password: String = rand::thread_rng()
+                .sample_iter(&Alphanumeric)
+                .take(15)
+                .map(char::from)
+                .collect();
+
+            let mut secret_data = BTreeMap::new();
+            secret_data.insert("nifiSensitivePropsKey".to_string(), password);
+
+            let new_secret = Secret {
+                metadata: ObjectMetaBuilder::new()
+                    .namespace(namespace)
+                    .name(&sensitive_config.key_secret.to_string())
+                    .build(),
+                string_data: Some(secret_data),
+                ..Secret::default()
+            };
+            client
+                .create(&new_secret)
+                .await
+                .with_context(|_| SensitiveKeySecretSnafu {})?;
+            Ok(true)
+        }
+    }
+}
+
+fn get_stackable_secret_volume_attributes() -> BTreeMap<String, String> {
+    let mut result = BTreeMap::new();
+    result.insert(
+        "secrets.stackable.tech/class".to_string(),
+        "tls".to_string(),
+    );
+    result.insert(
+        "secrets.stackable.tech/scope".to_string(),
+        "node,pod".to_string(),
+    );
+    result
+}
+
+fn build_persistent_volume_claim_rwo_storage(name: &str, storage: &str) -> PersistentVolumeClaim {
+    PersistentVolumeClaim {
+        metadata: ObjectMeta {
+            name: Some(name.to_string()),
+            ..ObjectMeta::default()
+        },
+        spec: Some(PersistentVolumeClaimSpec {
+            access_modes: Some(vec!["ReadWriteOnce".to_string()]),
+            resources: Some(ResourceRequirements {
+                requests: Some({
+                    let mut map = BTreeMap::new();
+                    map.insert("storage".to_string(), Quantity(storage.to_string()));
+                    map
+                }),
+                ..ResourceRequirements::default()
+            }),
+            ..PersistentVolumeClaimSpec::default()
+        }),
+        ..PersistentVolumeClaim::default()
+    }
+}
+
+async fn external_node_port(nifi_service: &Service) -> Result<i32> {
+    let external_ports = nifi_service
+        .spec
+        .as_ref()
+        .with_context(|| ObjectHasNoSpecSnafu {})?
+        .ports
+        .as_ref()
+        .with_context(|| ExternalPortSnafu {})?
+        .iter()
+        .filter(|p| p.name == Some(HTTPS_PORT_NAME.to_string()))
+        .collect::<Vec<_>>();
+
+    let port = external_ports
+        .first()
+        .with_context(|| ExternalPortSnafu {})?;
+
+    Ok(port.node_port.with_context(|| ExternalPortSnafu {})?)
+}
+
+async fn get_proxy_hosts(
```

Comment:
Is there a reason you don't use the NodePort per Pod approach like other operators ? That way you don't need extra requests to the api server and you "just know" the names of your hosts. Doesn't NiFi support names instead of IPs?

Reply:
To be honest I simply inherited that part of the code from @maltesander .
Happy to change it to be in line with others operators of course.

The reason for this proxy hosts function is a bit more involved though, as this has to be set for people to be able to connect to NiFi from outside of Kubernetes.
NiFi checks http headers and by default rejects anything that was not sent to the host that it bound itself to.

So for a nodeport for example I address node-external-ip:32454 from my laptop, that is then forwarded to nifi-pod-1....:8443
Nifi now throws this away because "I'm not node-external-ip:32454". 
So we pretty much need to stick everything in that proxy setting that someone could presumably use to connect to NiFi.

Reply:
I've thought about this, and I think functionaly there is no difference between a nodeport that is configured to always forward to the local node and three nodeports with only that node configured?
Except that if we use three nodeports I think we actually have three different port numbers, this way we only have one for all nodes?